presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java

```diff
@@ -488,6 +504,46 @@ public void evaluateFinal(int groupId, BlockBuilder output)
         public void prepareFinal()
         {
         }
+
+        @Override
+        public GroupIdPage addInputPreprocessing(GroupIdPage groupIdPage)
+        {
+            GroupByIdBlock groupIdsBlock = groupIdPage.getGroupByIdBlock();
+
+            // Prepend the groupId block to the input page
+            Page withGroup = groupIdPage.getPage().prependColumn(groupIdsBlock);
+
+            // Filter out positions based on mask, if present
+            Page filtered;
+            if (maskChannel >= 0) {
+                filtered = filter(withGroup, withGroup.getBlock(maskChannel + 1)); // offset by one because of group id in column 0
+            }
+            else {
+                filtered = withGroup;
+            }
+
+            // Compute a mask for the distinct rows (including the group id)
+            Work<Block> work = hash.markDistinctRows(filtered);
+            checkState(work.process());
+            Block distinctMask = work.getResult();
+
+            // Filter out duplicate rows and return the page with distinct rows
+            Page dedupPage = filter(filtered, distinctMask);
+
+            // Get updated GroupByIdBlock from distinctPage
+            GroupByIdBlock distinctGroupIdsBlock = new GroupByIdBlock(groupIdsBlock.getGroupCount(), dedupPage.getBlock(0));
+
+            // Remove the groupId block from distinct Page
+            Page distinctPage = dedupPage.extractChannels(IntStream.range(1, dedupPage.getChannelCount()).toArray());
```

Comment:
Creating `IntStream.range(1, dedupPage.getChannelCount()).toArray()` for each input page seems unnecessarily expensive, I suggesting storing that array as a field to avoid the per call overhead.